src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+
+abstract class EvpKey implements Key, Destroyable {
+    private static final long serialVersionUID = 1;
+
+    protected final InternalKey internalKey;
+    protected final EvpKeyType type;
+    protected final boolean isPublicKey;
+    protected boolean ephemeral = false;
+
+    protected byte[] encoded;
+    protected Integer cachedHashCode;
+
+    private static native void releaseKey(long ptr);
+    private static native byte[] encodePublicKey(long ptr);
+    private static native byte[] encodePrivateKey(long ptr);
+    protected static native byte[] getDerEncodedParams(long ptr);
+
+    EvpKey(final InternalKey key, final EvpKeyType type, final boolean isPublicKey) {
+        this.internalKey = key;
+        this.type = type;
+        this.isPublicKey = isPublicKey;
+    }
+
+    boolean isEphemeral() {
+        return ephemeral;
+    }
+
+    void setEphemeral(final boolean ephemeral) {
+        this.ephemeral = ephemeral;
+    }
+
+    void releaseEphemeral() {
+        if (ephemeral) {
+            destroy();
+        }
+    }
+
+    // @CheckReturnValue // Restore once replacement for JSR-305 available
+    <T, X extends Throwable> T use(final MiscInterfaces.ThrowingLongFunction<T, X> function) throws X {
+        return internalKey.use(function);
+    }
+
+    <X extends Throwable> void useVoid(final MiscInterfaces.ThrowingLongConsumer<X> function) throws X {
+        internalKey.useVoid(function);
+    }
+
+    @Override
+    public String getAlgorithm() {
+        return type.jceName;
+    }
+
+    @Override
+    public String getFormat() {
+        return isPublicKey ? "X.509" :  "PKCS#8";
+    }
+
+    @Override
+    public byte[] getEncoded() {
+        initEncoded();
+        return encoded != null ? encoded.clone() : encoded;
+    }
+
+    protected synchronized void initEncoded() {
+        if (encoded == null) {
+            encoded = isPublicKey ? use(EvpKey::encodePublicKey) : use(EvpKey::encodePrivateKey);
+        }
+    }
+
+    protected <X extends Throwable> BigInteger nativeBN(final MiscInterfaces.ThrowingLongFunction<byte[], X> fn)
+            throws X {
+        byte[] raw = use(fn::apply);
+        return new BigInteger(1, raw);
+    }
+
+    protected <T extends AlgorithmParameterSpec> T nativeParams(final Class<T> paramSpec) {
+        byte[] encodedParams = use(EvpKey::getDerEncodedParams);
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance(type.jceName);
+            params.init(encodedParams);
+            return params.getParameterSpec(paramSpec);
+        } catch (final GeneralSecurityException | IOException ex) {
+            throw new RuntimeCryptoException("Unable to deserialize parameters: "
+                + Base64.getEncoder().encodeToString(encoded), ex);
+        }
+    }
+
+    /**
+     * This method will be called by @{link #destroy()} after calling @{code internalKey.release()}.
+     */
+    protected synchronized void destroyJavaState() {
```

Comment:
seems like the tradeoff here is between performance and future-proofing. if we declare it `synchronized` on the parent object, subclasses that may not need synchronization will inherit that behavior regardless. however, if we don't synchronize the base method, we rely on the best intentions of future programmers to always know whether they need to `synchronize` a subclass's override.

arguing from intuition here, but i would bet that the performance to be gained from not `synchronize`'ing no-op impl.'s of this method would be marginal because it seems like most (if not all) `destroyJavaState` impls would be cheap and thus not block for long. however, the cost of accidentally letting a race condition creep into this teardown logic seems like a non-negligible correctness risk.

so, i suppose i'm on board with leaving it as-is, `synchronize`d on the base method.